pkg/vtpm/src/server.cpp

```diff
@@ -33,65 +46,350 @@ google::protobuf::uint32 readHdr(char *buf)
+//is given as input.
+//Returns non-zero error on failure.
+int
+handleRequest (int sock, google::protobuf::uint32 size)
+{
+    char payload[size+CODED_STRM_HDR_LEN];
+    eve_tools::EveTPMRequest request;
+    eve_tools::EveTPMResponse response;
+    int byteCnt = 0, rc = success;
+    const char *cmdArgs[MAX_ARGS];
+    ifstream cmdOut;
+
+    if (parseRequest(sock, size, request, response, payload) < 0) {
+        sendResponse(sock, response);
+        rc = failure;
+        goto cleanup_and_exit;
+    }
+
+    if (sanitizeCmdRequest(sock, request, response) != 0) {
+        //send the error set by sanitizeCmdRequest to client.
+        sendResponse(sock, response);
+        rc = failure;
+        goto cleanup_and_exit;
+    }
+
+    if (prepareCommand(request.command(), cmdArgs, response) < 0) {
+        sendResponse(sock, response);
+        rc = failure;
+        goto cleanup_and_exit;
+    }
+
+    //Prepare input files expected by the command.
     for (int i=0; i < request.inputfiles_size(); i++) {
         const eve_tools::File& file = request.inputfiles(i);
-        cout << "Processing file: " << file.name() << std::endl;
         ofstream input_file;
         input_file.open(file.name(), ios::out|ios::binary);
         if (!input_file) {
-            cout << "Unable to open test file for writing" << std::endl;
-            return -1;
+            cerr << "Unable to open input file for writing: "
+                 << file.name() << std::endl;
+            response.set_response("Unable to open input file " + file.name());
+            sendResponse(sock, response);
+            rc = failure;
+            goto cleanup_and_exit;
         }
         input_file << file.content();
         input_file.close();
     }
 
-    ostringstream command;
-    command << request.command() << " " << "> cmd.output 2>&1";
-    system(command.str().c_str());
-    eve_tools::EveTPMResponse response;
-    ifstream cmdOut;
-    cmdOut.open("cmd.output", ios::in);
-    if (!cmdOut) {
-        cout << "Error opening cmd.output, skipping it" << std::endl;
-    } else {
+    //sync all the input files.
+    sync();
+
+    rc = execCmd(cmdArgs);
+    if (rc != 0) {
+        cerr << "Command invocation failed with rc " << rc << std::endl;
+        response.set_response("Backend failure while serving the request");
+        sendResponse(sock, response);
+        rc = failure;
+        goto cleanup_and_exit;
+    }
+
+    //sync all the output files.
+    sync();
+
+    //Pack stderr/stdout from command invocation.
+    cmdOut.open(CMD_OUTPUT_FILE, ios::in);
+    if (cmdOut) {
         ostringstream cmdoutstream;
         cmdoutstream << cmdOut.rdbuf();
+        cout << "Command output is: " << std::endl
+             << cmdoutstream.str() << std::endl;
         response.set_response(cmdoutstream.str());
-        cout << "Command output is: " << std::endl;
-        cout << cmdoutstream.str() << std::endl;
+        cmdOut.close();
     }
```

Comment:
Is it an error if cmdOut is not there i.e., the command was expected to produce some output but did not?